_posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md

```diff
@@ -0,0 +1,290 @@
+            concatenate_sources => true
+            key => "any-long-encryption-key"
+            method => "SHA1"    # return the same hash if all values of source fields are equal
+        }
+
+    }
+}
+{% endraw %}{% endhighlight %}
+
+In the listing above you can see an excerpt of the filter part of the configuration file. The first statement adds a new field `report_url` to the JSON object. Therefore, we concatenate an environment variable with a field that was defined in the original JSON object to obtain a complete URL. The second statement creates a fingerprint needed for the Elasticsearch document ID that will be added to a metadata field. The fingerprint will be created from the fields specified by the key `source`.
+
+{% highlight text %}{% raw %}
+{%- if "elasticsearch" in LS_OUTPUT or "document" in LS_OUTPUT or "template" in LS_OUTPUT %}
+
+############################
+# Output for elasticsearch #
+############################
+
+# only if no error tags were created
+if (![tags]) {
+
+    # push esf events to elasticsearch
+    elasticsearch {
+
+       # set connection
+       hosts => {{ ES_HOSTS }}
+
+       {%- if ES_USER and ES_PASSWORD %}
+
+       # set credentials
+       user => "{{ ES_USER }}"
+       password => "{{ ES_PASSWORD }}"
+       {%- endif %}
+
+       {%- if "elasticsearch" in LS_OUTPUT or "document" in LS_OUTPUT %}
+
+       # set document path
+       index => "{{ ES_INDEX }}"
+       document_type => "{{ ES_DOCUMENT_TYPE }}"
+       document_id => "%{[@metadata][ES_DOCUMENT_ID]}"
+       {%- endif %}
+
+       {%- if "elasticsearch" in LS_OUTPUT or "template" in LS_OUTPUT %}
+
+       # use template for settings and mappings
+       manage_template => true
+       template => "{{ LS_CONFIG_VOL }}/template-esf.json"
+       template_name => "{{ ES_INDEX }}"
+       template_overwrite => true
+       {%- endif %}
+    }
+}
+{%- endif %}
+{% endraw %}{% endhighlight %}
+
+
+The above code snippet shows how we push the output to our Elasticsearch cluster. The first line represents how we use our own templating engine. If the `if`-statement is `false`, the part configuring the output to Elasticsearch will be omitted. If it equals `true`, we use environment variables as well as information from our metadata for connection, document path and Elasticsearch index template, which contains index settings and the mappings of document fields to the desired data types. We also have written a similar construct to generate the output to stdout, an error log file and an info log file, which later contains successfully pushed data for monitoring reasons.
+
+#### Testing
+
+The CI part for our Logstash container is very similiar to that of the Elasticsearch container. The tests we run are different of course, but the process is the same.
+
+### Part 4 - Integrate Docker Containers in Pipeline with Jenkins
```

Comment:
Integrate Docker Containers in Pipeline with Jenkins >  Integrate Docker containers in pipeline with Jenkins